cpp/src/reductions/scan.cu

```diff
@@ -0,0 +1,163 @@
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/iterator.cuh>
+
+#include <rmm/rmm.h>
+#include <cudf/utilities/error.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <utilities/device_atomics.cuh>
+#include <cub/device/device_scan.cuh>
+#include <cudf/reduction.hpp>
+
+
+namespace cudf {
+namespace experimental {
+
+namespace detail {
+
+template <typename T, typename Op>
+struct Scan {
+    static
+    void call(const column_view input, mutable_column_view output,
+                  bool inclusive, cudaStream_t stream,
+                  rmm::mr::device_memory_resource* mr)
+    {
+        size_t size = input.size();
+
+        // Prepare temp storage
+        void *temp_storage = NULL;
+        size_t temp_storage_bytes = 0;
+
+        if (input.has_nulls()) {
+          auto d_input = column_device_view::create(input, stream);
+          auto it = make_null_replacement_iterator(*d_input, Op::template identity<T>());
+          auto scan_function = (inclusive ? inclusive_scan<decltype(it)> : exclusive_scan<decltype(it)>);
+          scan_function(temp_storage, temp_storage_bytes,
+              it, output.data<T>(), size, stream);
+          rmm::device_buffer buff(temp_storage_bytes, stream, mr);
+          temp_storage = buff.data(); 
+
+          scan_function(temp_storage, temp_storage_bytes,
+              it, output.data<T>(), size, stream);
+        } else {
+          auto it = input.data<T>();  //since scan is for arithmetic types only
+          auto scan_function = (inclusive ? inclusive_scan<decltype(it)> : exclusive_scan<decltype(it)>);
+          scan_function(temp_storage, temp_storage_bytes,
+              it, output.data<T>(), size, stream);
+          rmm::device_buffer buff(temp_storage_bytes, stream, mr);
+          temp_storage = buff.data(); 
+
+          scan_function(temp_storage, temp_storage_bytes,
+              it, output.data<T>(), size, stream);
+        }
+    }
+
+    template <typename InputIterator>
+    static void exclusive_scan(void *&temp_storage, size_t &temp_storage_bytes,
```

Comment:
Just make these take the column_view / mutable_column_view and allocate the temp storage inside the function. No reason to have raw pointers in the interface.

Reply:
allocated temporary storage inside, but tparam `InputIterator` is needed because of `make_null_replacement_iterator` and `column_view::data<T>()`